### DISCUSSION TITLE
Pybricks for NXT

### DISCUSSION THREAD

**Discussion by skippa** [pybricks/support#939] [General]
Hi folks, just wondering if anyone got pybricks to work with the NXT brick? I read the nxos repo, I can probably hack my way to build a firmware, but I don't want to "brick my brick" so I was hoping someone had done it and could give me some help...

I got nxt-python to work a few years ago, but it's buggy and slow, really want to get this to work!

**Comment by dlech** [comment]
It is currently possible to boot Pybricks on NXT, but there is still quite a bit of work to do to make if fully functional.  If you want to try it, you can grab the [latest build from CI ](https://github.com/pybricks/pybricks-micropython/suites/10745069382/artifacts/539132296), use  a paper clip to press the hidden reset button on the NXT to put it in bootloader mode and flash it with `pipx run pybricksdev flash nxt-firmware-build-2436.zip`.

You don't need to worry about bricking the NXT. There is a permanent bootloader on the NXT, so even if you botch the firmware install, you can always activate the bootloader using the reset button on the NXT brick.

There is also an open issue for full support for NXT at #122. 


Also, FYI, the Pybricks NxOS fork is now merged into https://github.com/pybricks/pybricks-micropython so we won't be updating https://github.com/pybricks/nxos anymore.

**Reply by BertLindeman** [reply]
> It is currently possible to boot Pybricks on NXT, but there is still quite a bit of work to do to make if fully functional. If you want to try it, you can grab the [latest build from CI ](https://github.com/pybricks/pybricks-micropython/suites/10745069382/artifacts/539132296), use a paper clip to press the hidden reset button on the NXT to put it in bootloader mode and flash it with `pipx run pybricksdev flash nxt-firmware-build-2436.zip`.
> 
> You don't need to worry about bricking the NXT. There is a permanent bootloader on the NXT, so even if you botch the firmware install, you can always activate the bootloader using the reset button on the NXT brick.
> 
> There is also an open issue for full support for NXT at #122.
> 
> Also, FYI, the Pybricks NxOS fork is now merged into https://github.com/pybricks/pybricks-micropython so we won't be updating https://github.com/pybricks/nxos anymore.

David, this is an old post, so the links do not all work 
A link to [the latest CI build might be](https://github.com/pybricks/pybricks-micropython/actions/workflows/build.yml)

But I wonder if the current firmware for the NXT indeed provides a running nxos. 
I can pair and use an rfcomm bind to connect to the NXT. 
Using the terminal you link above or minicom or python I can make a serial connection and send a command, but the response on ```help()```  I get is: ```Error during communication: device reports readiness to read but returned no data (device disconnected or multiple access on port?)``` 

Or should I shut-up `there are more important things to do` and right you are 😃 

Just because there was new interest lately I wanted to see if I could get the REPL working.

Bert

**Reply by laurensvalk** [reply]
While the REPL does work locally on my machine, I wouldn't attach too much value to it. Its only purpose currently is to try out new developments as we make them. For example, I've been working on motors (not pushed yet) and the REPL was helpful to try motor commands. But the REPL isn't implemented in its final form at all. `stdio` is currently implemented by a few blocking functions in `mphalport.c` but need to be replaced by the proper protocol in `pbsys` eventually.

We're also actively looking for ways to support this initiative, and we may be making some small progress. Nothing final to report yet though.

**Reply by BertLindeman** [reply]
Thank you Laurens. 
Just curious if I could help, but you know I am no programmer (not really).

**Reply by laurensvalk** [reply]
By the way, the expected output/procedure is currently as follows:
- Load the firmware. Reboot the hub (to work around the secondary controller not starting after updating).
- Connect in your OS to the serial port. On Ubuntu for example, use Blueman and right-click to `connect to serial Port`.
- You may need to confirm pin: `1234`
- Open your terminal emulator like `screen` or `picocom` to open to the port. (` /dev/rfcomm0` for me)
- Type one symbol. You should get a hint about this on the NXT screen. When accepted, it will indicate `REPL`.  This is needed to get `stdio` going, due to not-so-great implemetation of `stdio` as indicated above.
- Then your terminal will show the output of the following [baked in script](https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/drv/block_device/block_device_test.c): So if you have no way of showing output, it may appear that input is ignored for 10 seconds.

```python
from pybricks.tools import StopWatch, wait

timer = StopWatch()

for i in range(10):
    print("Hello, world at time (ms):", timer.time())
    wait(1000)

raise KeyboardInterrupt
```
- Then it drops to the REPL like it always does with `raise KeyboardInterrupt`. You can't skip this with `CTRL+C` by the way.

**Reply by laurensvalk** [reply]
> Thank you Laurens. Just curious if I could help, but you know I am no programmer (not really).

@BertLindeman, you are super helpful as it is :smile: Look what I got the other day:

> It was quite a challenge for the primary school students I teach to make the switch from Spike to Pybrics.
At a critical moment when we really didn't know how to proceed due to technical problems **Bert Lindeman** and **David Lechner** jumped in with the answer (on Github).
>
> And then the great project that my students had just had to be implemented through the excellent Pybrics firmware.
>
> The SILVER medal won the team I teach STEM........and the rest is history...

So thank you!

**Reply by BertLindeman** [reply]
Getting all warm. 
Thank you for sharing this.
Will go back and retry the NXT as it **is** a fun device.
And try be not too talkative.

**Reply by laurensvalk** [reply]
I will try to push some updates some time soonish. It's possible that I'm building off an older branch and something broke more recently. I do see a potential issue in the latest build :wink:

**Reply by BertLindeman** [reply]
No problem Laurens.

**Reply by BertLindeman** [reply]
Confirm getting the REPL flashing the oldest nxt firmware I have: `nxt-firmware-build-3357-gitd3f38000.zip`

```
Hello!
World!
Traceback (most recent call last):
  File "hello.py", line 8, in <module>
KeyboardInterrupt:
Pybricks MicroPython ci-build-3357-v3.5.0b1-12-gd3f38000 on 2024-04-04; LEGO MINDSTORMS NXT Brick with AT91SAM7S256
Type "help()" for more information.
>>>
```

For my memory:
Flash with ```pipx run pybricksdev  nxt-firmware-build-3357-gitd3f38000.zip'```
The device was already paired.
Setup serial interface: ```sudo rfcomm release /dev/rfcomm1 ; sudo rfcomm bind /dev/rfcomm1 00:16:53:0C:2E:97 ```
get terminal: ```screen /dev/rfcomm1```

FUN!

**Reply by laurensvalk** [reply]
Nice to see you here! Latest links are here: https://nightly.link/pybricks/pybricks-micropython/workflows/build/master

**Comment by skippa** [comment]
Excellent! Thanks for your reply, I'll have a go and see what it can do,
then report back

**Reply by skippa** [reply]
Ok, so I was able to successfully install the firmware on my NXT.  I could also pair the bluetooth device.  However, I cannot connect to bluetooth in pybricks.  I'm using Chrome and have enabled web bluetooth as per the instructions... Anything else I can try?

**Reply by laurensvalk** [reply]
That part of it is not implemented yet, and probably won't be for a while :)

But if you've paired with your computer, you should be able to open the virtual com port that is now available. On Linux, you could use `screen`, or `Putty` on Windows, for example.

**Reply by dlech** [reply]
Or https://googlechromelabs.github.io/serial-terminal on any OS.

**Reply by Leolelion2112** [reply]
> Ok, so I was able to successfully install the firmware on my NXT. I could also pair the bluetooth device. However, I cannot connect to bluetooth in pybricks. I'm using Chrome and have enabled web bluetooth as per the instructions... Anything else I can try?

I can't seem to pair to the NXT. It does connect for a few seconds, then it goes out.

**Reply by Leolelion2112** [reply]
I get "Connection failed - Other Device".

**Reply by dlech** [reply]
You need to use the COM port created for it rather than trying to connect manually. The OS will connect automatically once you open the COM port.

**Comment by skippa** [comment]
Ah ok, I'll have a play

How complicated is this to implement? Maybe I could assist...

**Reply by dlech** [reply]
1. The [Pybricks Profile specification](https://github.com/pybricks/technical-info/blob/master/pybricks-ble-profile.md) needs to be modified to allow all functionality over an RFCOMM socket.
    1. Instead of a separate NUS service, we need to add a new command and a new event to the Pybricks service for stdio.
    2. We also need a command(s) and event(s) for the Device Information Service.
    3. Since RFCOMM is essentially a stream of bytes, we will need to add a header containing the message size and sequence number to encapsulate Pybricks Service commands and events in the stream.
    4. We need a new event type that can act as the response to commands that are "write with response" in the GATT version of the protocol.
2. Alternately, we could do something that approximates GATT over RFCOMM.
    1.  This would need the same size and sequence number header but also include a GATT command/event and attribute handle
    2. The reset of the message would contain the same data as the current Pybricks Profile dictates
3. One the protocol is worked out on paper, it needs to be implemented in the firmware, in pybricksdev and in Pybricks Code.

**Comment by skippa** [comment]
Any update or progress on this?

**Reply by dlech** [reply]
We have defined the Pybricks Profile over USB in https://github.com/pybricks/pybricks-micropython/pull/208. It isn't quite finalized yet, but should be very close.

If you are still interested in helping, porting that to NXT would be a good place to start.

**Comment by damienmckenna** [comment]
It's a year later and I saw pybricks/pybricks-micropython#208 was merged, what's next to get this to work? Thank you.

**Reply by laurensvalk** [reply]
We posted this [call for support](https://fosstodon.org/@laurensvalk/114743346787502504) a few weeks ago and there have been some very exciting code contributions lately, so this is definitely gathering some steam.

We've been collecting some [issues here](https://github.com/pybricks/support/issues?q=is%3Aissue%20sort%3Aupdated-desc%20label%3A%22help%20wanted%22) that could use some help. We should probably open some more now that there is so much going on :smile: 

Also tagging OP @skippa.

**Comment by Leolelion2112** [comment]
Hello? I'd like to have some answers for NXT Pybricks:
- Can you change the hub name?
- Could i help add things for the NXT on the docs?
- Can programs be saved on the NXT?
- Is Bluetooth available on the NXT?

**Reply by Leolelion2112** [reply]
Hello? Is anyone online today?

**Reply by dlech** [reply]
NXT is still very much a work-in-progress. So most of these things are planned, but not working yet. So it is a little early to be starting to work on the documentation.

**Reply by Leolelion2112** [reply]
I mean i saw a lot of things while in the serial console and wanted to know if i could add a bit of documentation just so i (and new people to this) don't get lost..

**Reply by Leolelion2112** [reply]
> I mean i saw a lot of things while in the serial console and wanted to know if i could add a bit of documentation just so i (and new people to this) don't get lost..

Hello? I wonder if anyone is here?

**Reply by laurensvalk** [reply]
As we've pointed out on the Patreon post, we appreciate your enthusiasm but unfortunately there isn't much to try or document for NXT yet. Posting `Hello?` 7 times or deleting it, and posting `Hello?` again does not necessarily make it appear any faster :smile:

**Reply by Leolelion2112** [reply]
Alright, sorry. But thanks for the response!

**Reply by Leolelion2112** [reply]
Hi, just a small question: when i try loading in NXT sensors in NXT Pybricks (sorry, i'm a bit excited about it), i get error 95. Is it because S1-4 ports are still uncontrollable?

**Reply by jaguilar** [reply]
FWIW, I found the code relatively accessible. It might be best to just fire up that github search function and start looking for where that error code might be produced. 

I don't see obvious candidates for error 95 when I search, so I would instead start looking for where in the code the action you tried is happening. 

When you say "load NXT sensors in NXT pybricks", I'm assuming you were doing something like instantiating a touch sensor. The code that instantiates that sensor is [here](https://github.com/pybricks/pybricks-micropython/blob/master/pybricks/nxtdevices/pb_type_nxtdevices_touchsensor.c). You could start tracing through that code to see where it gets into unimplemented stuff. Or you could start wiring it up with pbdrv_uart_debug_printf calls to see if you can trace down what's happening. Tracing through the code like that will hopefully also give you a sense of what's done and not done.

That being said, I'm not sure giving feedback like, "I'm getting error 95" is really going to help much, because the fellas already know that NXT isn't working yet. So this would only be pointful if your plan was to dive in and start writing code to get things working.

**Reply by Leolelion2112** [reply]
I mean when i try to init any sensor on the S ports i get:
>>> woow = SoundSensor(Port.S2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 1] EPERM: 

The requested operation is not valid in the current state:
--> Check the documentation for required conditions.
--> Check the line in your script that matches
    the line number given in the 'Traceback' above.

As i do have a Sound Sensor on Port S2 of the NXT, i do not understand why is.

**Reply by laurensvalk** [reply]
Sensors are not yet implemented on NXT unfortunately, so they will return errors when trying to use them.

We're much further along than I thought we'd ever be though, so we might actually get there at some point.

For the next few months at least, EV3 and Spike have our priority, but we're hoping to get around to working on NXT sooner than later.

**Reply by Leolelion2112** [reply]
Ok! I hope sensor support comes out soon! Can't wait!

**Reply by Leolelion2112** [reply]
Found out about build #4534 in pybricks-micropython's actions and saw "partial support for nxt sensors". When i flashed it, it gave me its MAC address on Bluetooth and randomly started ticking. I can't connect to it either. User error?

**Reply by jaguilar** [reply]
Unless you're willing to self-diagnose issues, it might be best to wait until a beta release where NXT support is announced, or until someone explicitly asks you to try a build.

**Reply by Leolelion2112** [reply]
My NXT's not connecting over bluetooth in the latest release. Could anyone fix this?

**Reply by laurensvalk** [reply]
Bluetooth has been disabled because USB will replace it. In a future release, we'll try to include NXT.

**Reply by Leolelion2112** [reply]
Can you implement program download too if possible?
